x/ratesync/types/types.go

```diff
@@ -0,0 +1,211 @@
+	}
+
+	if hc.IcaAccount.Owner != "" {
+		portID, err := icatypes.NewControllerPortID(hc.IcaAccount.Owner)
+		if err != nil {
+			return err
+		}
+		err = host.PortIdentifierValidator(portID)
+		if err != nil {
+			return err
+		}
+	}
+
+	switch hc.IcaAccount.ChannelState {
+	case liquidstakeibctypes.ICAAccount_ICA_CHANNEL_CREATING:
+		if hc.IcaAccount.Address != "" {
+			return fmt.Errorf("ica account address for ICAAccount_ICA_CHANNEL_CREATING should be empty")
+		}
+		// No features allowed without ICA account.
+		if hc.Features.LiquidStake.Enabled == true || hc.Features.LiquidStakeIBC.Enabled == true {
+			return fmt.Errorf("no features should be enabled without a valid ICA account")
+		}
+	case liquidstakeibctypes.ICAAccount_ICA_CHANNEL_CREATED:
+		if hc.IcaAccount.Address == "" {
+			return fmt.Errorf("ica account address for ICAAccount_ICA_CHANNEL_CREATED should not be empty")
+		}
+		_, _, err = bech32.DecodeAndConvert(hc.IcaAccount.Address)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = hc.Features.ValdidateBasic()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (hc HostChain) IsActive() bool {
+	if hc.Features.LiquidStakeIBC.Enabled == true ||
+		hc.Features.LiquidStake.Enabled == true {
+		return true
+	}
+	return false
+}
+
+func (f Feature) ValdidateBasic() error {
+	if f.LiquidStakeIBC.FeatureType != FeatureType_LIQUID_STAKE_IBC {
+		return fmt.Errorf("invalid feature type expected %s, got %s", FeatureType_LIQUID_STAKE_IBC, f.LiquidStakeIBC.FeatureType)
+	}
+	err := f.LiquidStakeIBC.ValdidateBasic()
+	if err != nil {
+		return err
+	}
+
+	if f.LiquidStake.FeatureType != FeatureType_LIQUID_STAKE {
+		return fmt.Errorf("invalid feature type expected %s, got %s", FeatureType_LIQUID_STAKE, f.LiquidStake.FeatureType)
+	}
+	err = f.LiquidStake.ValdidateBasic()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (lsConfig LiquidStake) ValdidateBasic() error {
+	if lsConfig.CodeID == 0 {
+		if lsConfig.Instantiation != InstantiationState_INSTANTIATION_NOT_INITIATED {
+			return fmt.Errorf("config with 0 code id should not have been initiated")
+		}
+	}
+	switch lsConfig.Instantiation {
+	case InstantiationState_INSTANTIATION_NOT_INITIATED:
+		if lsConfig.ContractAddress != "" {
+			return fmt.Errorf("InstantiationState_INSTANTIATION_NOT_INITIATED cannot have contract address")
+		}
+		if lsConfig.Enabled == true {
+			return fmt.Errorf("feature cannot be turned on without instantiation complete")
+		}
+	case InstantiationState_INSTANTIATION_INITIATED:
+		if lsConfig.ContractAddress != "" {
+			return fmt.Errorf("InstantiationState_INSTANTIATION_INITIATED cannot have contract address")
+		}
+		if lsConfig.CodeID == 0 {
+			return fmt.Errorf("InstantiationState_INSTANTIATION_INITIATED cannot have 0 codeID")
+		}
+		if lsConfig.Enabled == true {
+			return fmt.Errorf("feature cannot be turned on without instantiation complete")
+		}
+	case InstantiationState_INSTANTIATION_COMPLETED:
```

Comment:
check for duplicates here as well

Reply:
cannot(no keeper) and not needed, duplicates are allowed.

Reply:
can allow re-reg of same host-chain.. portID will be different, new instantiation by default ( as admin won't be changed)